File: composite-exo-glove/python/src/lifecycle.py
from __future__ import annotations

import argparse

from qnbot_sdk import CompositeExoGloveConfig, Sdk, SerialConnection
from qnbot_sdk.exo import ExoConfig
from qnbot_sdk.glove import GloveConfig

EXO_NAME = "exo"
GLOVE_NAME = "glove"


def create_sdk(port: str) -> Sdk:
    return Sdk(
        devices=(
            CompositeExoGloveConfig(
                connection=SerialConnection(port=port),
                devices=(GloveConfig(name=GLOVE_NAME), ExoConfig(name=EXO_NAME)),
            ),
        )
    )


def main() -> None:
    arguments = argparse.ArgumentParser(description="Run a short composite lifecycle")
    arguments.add_argument("--port", required=True, help="Shared serial port")
    options = arguments.parse_args()

    sdk = create_sdk(port=options.port)
    try:
        glove_domain = sdk.glove()
        exo_domain = sdk.exo()
        glove = glove_domain.device(GLOVE_NAME)
        exo = exo_domain.device(EXO_NAME)
        glove_domain.start()
        exo_domain.start()
        glove_domain.update()
        exo_domain.update()
        print(f"started Glove={glove.source_id} Exo={exo.source_id}")
        print(f"Glove pose={glove.pose().latest()}")
        print(f"Exo telemetry={exo.telemetry().latest()}")
        glove_domain.stop()
        exo_domain.stop()
        print("stopped")
    finally:
        glove_domain.close()
        exo_domain.close()
        print("closed")


if __name__ == "__main__":
    main()

File: composite-exo-glove/cpp/src/lifecycle.cpp
#include "composite_example_support.hpp"

#include <cstdlib>
#include <exception>
#include <iostream>

int main(int argc, char** argv) {
    try {
        qnbot::Sdk sdk(
            example::composite_config(example::parse_port(argc, argv)));
        auto glove_domain = sdk.glove();
        auto exo_domain = sdk.exo();
        auto glove = glove_domain.device("glove");
        auto exo = exo_domain.device("exo");
        glove_domain.start();
        exo_domain.start();
        static_cast<void>(glove_domain.update());
        static_cast<void>(exo_domain.update());
        const auto pose = glove.pose().latest();
        const auto telemetry = exo.telemetry().latest();
        std::cout << "started composite device; glove_pose="
                  << (pose ? "available" : "waiting")
                  << " exo_telemetry=" << (telemetry ? "available" : "waiting")
                  << '\n';
        glove_domain.stop();
        exo_domain.stop();
        std::cout << "stopped composite device\n";
        glove_domain.close();
        exo_domain.close();
        std::cout << "closed member domains\n";
        return EXIT_SUCCESS;
    } catch (const std::exception& error) {
        return example::report_error(error);
    }
}
